[ConstraintSystem] Overhaul ambiguity diagnostics

xedin opened the pull request:
- Don't attempt to rank solutions based on overloads if they have fixes
- Factor out overload ambiguity diagnostics
- Switch to a different algorithm:
  - Aggregate all of the available fixes based on callee locator;
  - For each ambiguous overload match aggregated fixes and diagnose;
  - Discard all of the fixes which have been already considered
    as part of overload diagnostics;
  - Diagnose remaining (uniqued based on kind + locator) fixes
     iff they appear in all of the solutions.

Resolves: [SR-12689](https://bugs.swift.org/browse/SR-12689)
Resolves: rdar://problem/62481592

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintSystem] Rank solutions based on overload choices only after fix diagnostics have failed

If there are multiple solutions with fixes, let's not try to rank
them based on overload choices because doing so would filter out
viable candidates.

xedin committed:
[ConstraintSystem] Encapsulate a way to diagnose ambiguity for aggregate fixes

xedin committed:
[ConstraintSystem] Overhaul ambiguity diagnostics

Refactor `diagnoseAmbiguityWithFixes` as follows:

- Aggregate all of the available fixes based on callee locator;
- For each ambiguous overload match aggregated fixes and diagnose;
- Discard all of the fixes which have been already considered
  as part of overload diagnostics;
- Diagnose remaining (uniqued based on kind + locator) fixes
  iff they appear in all of the solutions.

xedin committed:
[ConstraintSystem] NFC: Leave notes on the algorithm used by `diagnoseAmbiguityWithFixes`

xedin commented:
/cc @LucianoPAlmeida

xedin commented:
This is still a WIP because we need to implement `diagnoseForAmbiguity` for some of the fixes e.g. `treat r-value as l-value` and do other work, but I'd like to get some initial feedback on this to make sure that everybody on the same page.

LucianoPAlmeida commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -3099,6 +3031,112 @@ bool ConstraintSystem::diagnoseAmbiguityWithFixes(
+    const auto *fix = entry.second;
+    auto *calleeLocator = solution.getCalleeLocator(fix->getLocator());
+
+    fixesByCallee[calleeLocator].push_back({&solution, fix});
```
Just curious, but at this point is possible for `getCalleeLocator` to return a `nullptr` if it does what happens to the fix and solution?

hamishknight replied:
`getCalleeLocator` should never return `nullptr`, its default behaviour (if no other case applies) is to return a locator with the same anchor as the input locator.

LucianoPAlmeida commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -3099,6 +3031,112 @@ bool ConstraintSystem::diagnoseAmbiguityWithFixes(
+    auto aggregate = fixes->second;
+    diagnosed |= ::diagnoseAmbiguity(*this, ambiguity, aggregate, solutions);
+
+    consideredFixes.insert(aggregate.begin(), aggregate.end());
```
I wonder if In the cases where `diagnoseAmbiguity` returns false(if there's any) ... since we are adding them independently it maybe is a problem since they would be dropped from `fixes` undiagnosed? If there are such cases, maybe we should let them to attempt diagnose by kind?

xedin replied:
The reason why I did that is related to the fact that it could be a bunch of different fixes all related to the same callee e.g. `1 + ""` would produce one fix for first argument and one for second. Even if we didn't produce ambiguity related to overload in such case keeping them around wouldn't help to diagnose anything later because there is going to be no intersection so it's better to produce "I don't know" diagnostic and figure out case by case what went wrong...

LucianoPAlmeida replied:
Ah makes sense, I think I actually have seen situations like that too in SR-12689. I was just worried that it may end up discarding useful fixes in some way... but now I see the point :)

hborla reviewed (commented):
The new approach looks great!

xedin committed:
[TypeChecker] NFC: Adjust tests improved by new approach for ambiguity diagnosis

xedin committed:
[CSFix] Make it possible to diagnose r-value -> l-value fix for ambiguity

xedin committed:
[ConstraintSystem] Increase impact of fixes for some conditional requirements

If there is a conditional requirement failure associated with
member/function reference used in a call let's increase a score
of a fix for such failure because it renders member/function
unreachable in current context or with a given set of arguments.

xedin committed:
[ConstraintSystem] Specialize diagnostic for `~=` to talk about expression pattern use

xedin committed:
[ConstraintSystem] NFC/Debug: Log solutions before attempting to diagnose ambiguity with fixes

xedin committed:
[Diagnostics] If aggregate fix consists of identical fixes diagnose it specially

If aggregate fix (based on callee locator) differs only in picked
overload choices, let's diagnose that via `diagnoseForAmbiguity`
associated with a particular fix kind all solutions share, that
would produce a tailored diagnostic instead of the most general one.

xedin committed:
[Diagnostics] Check whether all contextual mismatches has the same type before diagnosing ambiguity

Instead of requiring sub-classes of `ContextualMismatch` to implement
`diagnoseForAmbiguity` let's implement it directly on `ContextualMismatch`
itself and check whether all of the aggregated fixes have same types on
both sides and if so, diagnose as-if it was a single fix.

xedin committed:
[Diagnostics] Diagnose ambiguity related to contextual type specifically

xedin committed:
[ConstraintSystem] Turn 'omit ' into a contextual mismatch it is

xedin committed:
[ConstraintSystem] Avoid recording duplicate fixes for contextual type mismatch

xedin committed:
[Diagnostics] Implement `IgnoreAssignmentDestinationType::diagnoseForAmbiguity`

For cases like this:

```swift
struct X {}
struct Y {}

func overloaded<T>(_ value: T) -> T { value }
func overloaded<T>(_ value: T) -> Int { 0 }

func test(x: inout X, y: Y) {
  x = overloaded(y)
}
```

Solver would record a `IgnoreAssignmentDestinationType` fix per overload,
`diagnoseForAmbiguity` could be used to properly diagnose ambiguity cases
like that.

xedin committed:
[ConstraintSystem] Increase impact of conformance failure associated with a known foundation type

Just like we already do for stdlib types, extend impact of missing
conformances on _known_ foundation types because it's unlikely to
be a valid suggestion to extend a foundation type with a new conformance.

xedin committed:
[Diagnostics] Allow property wrapper fixes to be diagnosed as ambiguity

xedin committed:
[Diagnostics] NFC: Adjust diagnostic test-cases improved by new ambiguity diagnosis

xedin committed:
[ConstraintSystem] Attach candidate notes either to decl (if it has valid loc) or AST node

xedin committed:
[Diagnostics] Add a tailored diagnostic for function mismatch in argument positions

If parameter has a function type and no argument overload choices
match let's diagnose that specifically.

Resolves: [SR-12689](https://bugs.swift.org/browse/SR-12689)
Resolves: rdar://problem/62481592

xedin committed:
[TypeChecker] NFC: Restore test-cases removed by incorrect merge

xedin commented:
This PR is finally ready for review. @LucianoPAlmeida this PR incorporates a fix for SR-12689 which means that it supersedes https://github.com/apple/swift/pull/31359

LucianoPAlmeida commented:
> This PR is finally ready for review. @LucianoPAlmeida this PR incorporates a fix for [SR-12689](https://bugs.swift.org/browse/SR-12689) which means that it supersedes #31359

Awesome @xedin \o/ 
Not sure if you've seen, but I left a comment on that PR few days ago with another SR that I believe may be fixed by these changes too, so maybe you can take a look if thats the case and add the regression test as well :)

xedin commented:
@LucianoPAlmeida I've seen it, thanks! I think it's a different problem those, we'd need to split `LabelingFailure` into one-per-mismatch, that way we could diagnose cases like that better...

LucianoPAlmeida commented:
> @LucianoPAlmeida I've seen it, thanks! I think it's a different problem those, we'd need to split `LabelingFailure` into one-per-mismatch, that way we could diagnose cases like that better...

Ahh I see, but that's fine ... thanks for verifying :))

LucianoPAlmeida commented on lib/Sema/CSFix.cpp:
```diff
@@ -252,6 +252,33 @@ bool ContextualMismatch::diagnose(const Solution &solution, bool asNote) const {
+                   etalonTypes.second->isEqual(types.second);
+          })) {
+    const auto &primary = commonFixes.front();
+    return primary.second->diagnose(*primary.first, /*aNote=*/false);
```
`/*aNote*/ ` -> `/*asNote*/` ?

xedin replied:
Done!

theblixguy commented on lib/Sema/CSFix.cpp:
```diff
@@ -252,6 +252,33 @@ bool ContextualMismatch::diagnose(const Solution &solution, bool asNote) const {
+            solution.simplifyType(fix->getToType())};
+  };
+
+  auto etalonTypes = getTypes(commonFixes.front());
```
What does “etalon” mean here?

xedin replied:
It just means that it's something everything else is compared against.

theblixguy replied:
Oh, TIL. I have never heard of this word before!

theblixguy commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -2814,6 +2814,9 @@ static void diagnoseOperatorAmbiguity(ConstraintSystem &cs,
                     operatorName.str());
         return;
       }
+    } else if (operatorName.is("~=")) {
```
It might be better to compare with `Id_MatchOperator`

xedin replied:
Done!

theblixguy commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -3062,114 +3065,139 @@ diagnoseAmbiguityWithEphemeralPointers(ConstraintSystem &cs,
-      // A common fix must appear in all solutions
-      auto &commonFixes = fixes.second;
-      if (commonFixes.size() != solutions.size()) continue;
+  auto &DE = cs.getASTContext().Diags;
```
Do you think it would be better to create a function like `getDiagnosticEngine` or `getDiags` which would do this internally, as I noticed it’s being done in many places.

xedin replied:
I think it's okay to just get it this way, because it's a straight reference to a property...

LucianoPAlmeida commented on test/expr/postfix/dot/init_ref_delegation.swift:
```diff
@@ -323,12 +323,12 @@ class TestOverloadSets {
   }
   
-  convenience init(a : Z0) { // expected-note{{candidate has partially matching parameter list (a: Z0)}}
+  convenience init(a : Z0) { // expected-note{{candidate expects value of type 'Z0' for parameter #1}}
```
From what I remember from the other PR, the calls that trigger those have also a relabel fix alongside the argument mismatch. I think this is fine, but do you think to mention that in some way could be valid in those situations?

xedin replied:
I don't really know, this is something I was thinking about as well. I don't really like `partially matching parameter list` note because it's as un-actionable as it gets but I don't know how to also mention a label here. It would be very unfortunate to produce two diagnostics/notes - one for labeling and one for type because that would be confusing...

LucianoPAlmeida replied:
Yeah, I agree that show two notes wouldn't be ideal. But in the other PR I also thought that `partially matching parameter list` was not ideal, but I did choose to show the relabel diagnostic, but to be honest there was no particular reason other than is better than `partially matching parameter list`. But in fact neither correct the type or relabel the call would fix the problem by itself because both type and label are wrong in this situation. 
I've been thinking about it and also not sure how to mention the label and also don't know how much we are loosing not showing one of them ... probably not much since as you mention both are actionable

xedin replied:
I think we could re-visit this once it's possible to produce a fix per label mismatch, that way we could probably coalesce type and label mismatch for each argument and produce a much better note...

xedin committed:
[Diagnostics] Use `Id_MatchOperator` to check whether given operator is pattern match

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test Linux platform
